### ISSUE TITLE
[Theming] Allow component specific global theme overrides

### ISSUE THREAD

**Issue by nadiia** [uber/baseweb#106]
There seems to be a need to override global styles per component due to design language differences and variations. Per component global overrides should be on a theme or the ThemeProvider level.

### Current Behavior
Currently, in order to override styles on a specific component across the whole application, you'd need to create and use a customized component that passes needed overrides to the BaseUI component.

### Potential Solutions
1. Let the `styled`function accept a third argument, that is a per component style override hook. A POC can be found [here](https://github.com/uber-web/baseui/commit/2bf7c99c3a2eba23682af1be44e9d0b9c590f28c).
   Some cons here:
   - Component overrides would be css declaration blocks or functions that return a css declaration block in order to support dynamic styles. As opposed to the theme that contains style values only and no css declarations.
2. Have optional component specific style values in the theme obj. [POC](https://github.com/uber-web/baseui/commit/411e18f7baf806b0c581063433d139292051f587).
   Cons:
   - The styled components code get's messy

**Comment by abmai** [comment]
Are we still looking for a solution here?

**Comment by schnerd** [comment]
Still seems like a problem worth solving, but I don't think we've settled on a solution yet. Folks should feel free to share feedback on the ones proposed above, or any new ideas they might have.

**Comment by nadiia** [comment]
There are some component specific props added to the theme creator. Those mainly handle colors for theme variations. 
As per more granular global component overrides, we've come to the decision that a current solution we'd recommend is to create a wrapper component for a Basse UI one that provides necessary overrides to it. And then import and reuse a wrapped customized component when needed.
```
// my-custom-avatar.js
import {mergeOverrides} from '@uber/baseui';
import {Autocomplete} from '@uber/baseui/avatar';

const avatarOverrides = {
      Root: {
        style: ({$theme}) => ({
          ...$theme.borders.border500,
        }),
      },
    };

export default function MyCustomAvatar(props) {
  const overrides = mergeOverrides(avatarOverrides, props.overrides);
  return <Avatar {...props} overrides={overrides} />;
}
```

**Comment by airhorns** [comment]
Totally get if the answer to this is a permanent no, but, I think it would be really awesome to be able to override components from the theme. Wrapper components accomplish the same thing for sure, but, they have a discoverability problem, especially for teams of any size. You have to know that a wrapper component exists to use it properly, and so if you're say new to a team or unfamiliar with all the nooks and crannies of a codebase, you might guess that just using the baseui component is the right thing to do when your team says "we use baseweb, it's great!". It's also nice to be able to copy code right from StackOverflow or the examples in the baseweb docs and have that be correct without this extra context about which ones you need to treat as special.

I can think of a few mitigations for discoverability like lint rules or re-exporting every baseui component and then ensuring only re-exported ones are used, but I don't think these are great. Adding lint rules slows down linting and they're actually kinda hard to get right IMO, and re-exporting all the components adds to the bundle size and makes end-users lives worse.

If the baseui overrides system pulled in component specific default overrides from the theme, I feel like we'd get the best of both worlds. Developers could collaborate on changing the defaults for components in one place where they already collaborate on changing the defaults for components (the theme), and using the example code from the baseui docs would inherit the right styling for your team and theme right from the place it gets the rest of the look and feel stuff (the theme). 

It'd also be really handy for folks who want swappable themes that make more invasive changes beyond just the theme properties, so that if you wanted to change the overrides between light and dark mode you could for example. Right now, you'd have to make all your wrapper components theme aware which is less than ideal.

Also, FWIW, I logged into the baseui slack to ask about this and found I was the third person in 3 weeks asking about doing exactly this from the theme, so it seems like it would be impactful and frequently help users of the framework.

Thanks for your consideration!

**Comment by tajo** [comment]
First time seeing this and like the idea. As long as the implementation doesn't make things messy which I think it should not. 
 not. Let's talk about it after holidays @chasestarr.

**Comment by airhorns** [comment]
Did you fine folks make any decisions around this? It would be super great to get for those of us who want to make themes that do fancier stuff!!

**Comment by tajo** [comment]
Yes, it's on our list but no specific timeline. Open to contribution though.

**Comment by tolgaouz** [comment]
Hey everyone, I'm working on a PR for this issue right now that implements @schnerd 's solution using the first case which is passing overrides along with theme to the ThemeProvider. I got it to work for some components but one down side is that I used `static contextType` on some of the components that are Class Objects. 

Is this something that should be avoided since that attribute can only be assigned to one context only? Because it was a pretty straight-forward approach for me. Just made some changes on the types and passed overrides with the ThemeContext .Provider then used them with `mergeOverrides` for each component on the initialization of overrides. This way I was able to use global overrides but also still have the option for _individual component overrides_ to **override** the _global overrides_.
Although this is a very trivial solution it doesn't really mess with the codebase the project has right now and just adds some extra lines.

Please let me know your thoughts, I can create the PR with commits and further explanation of everything if that would be for the better.

**Comment by tajo** [comment]
Is there a reason to not use `Context.Consumer` and `useContext` instead?

**Comment by tolgaouz** [comment]
You mean like wrapping class component with a functional one and using `useContext`, then exporting that as the default? Otherwise doing this on Class components would require wrapping render funcs with `Context.Consumer` and passing the value as props to children, then this would also require some more logic to be applied on children as well.

**Comment by tolgaouz** [comment]
Thanks for the suggestion @tajo! I created this HOC to wrap components while exporting them. 
```javascript
// helpers/withOverrides.js
export default function withOverrides(WrappedComponent:React.ComponentType<any>, componentName:String):React.ComponentType<any> {
  return class extends React.Component {
    static contextType = ThemeContext
    constructor(props) {
      super(props);
    }
    
    render() {
      // Grab forwardedRef from props and pass it down
      const {forwardedRef, ...rest} = this.props;
      // mergeOverrides from themeContext and props
      const overrides = mergeOverrides(this.context.overrides?.[componentName], this.props.overrides)
      return <WrappedComponent {...this.props} overrides={overrides} ref={forwardedRef} />;
    }
  };
  // Note the second param "ref" provided by React.forwardRef.
  // We can pass it along to withOverrides as a regular prop, e.g. "forwardedRef"
  // And it can then be attached to the Component.
  return React.forwardRef((props, ref) => {
    return <LogProps {...props} forwardedRef={ref} />;
  });
}
```
then simply, 

```javascript
// checkbox/checkbox.js
import withOverrides from '../helpers/withOverrides.js'
...
export default withOverrides(StatelessCheckbox,'Checkbox');
```
Now this component uses global overrides that are passed with the key `Checkbox`. I think this is a much cleaner approach then using static contextType directly in the component. What do you think about this approach?

**Comment by tajo** [comment]
Looks good conceptually to me - I like having an HOC like that. Not sure what this part is about tho:

```js
 // Note the second param "ref" provided by React.forwardRef.
  // We can pass it along to withOverrides as a regular prop, e.g. "forwardedRef"
  // And it can then be attached to the Component.
  return React.forwardRef((props, ref) => {
    return <LogProps {...props} forwardedRef={ref} />;
  });
```

**Comment by tolgaouz** [comment]
My bad! It's a snippet of code from React's official docs I forgot to change the name!

> Looks good conceptually to me - I like having an HOC like that. Not sure what this part is about tho:
> 
> ```js
>  // Note the second param "ref" provided by React.forwardRef.
>   // We can pass it along to withOverrides as a regular prop, e.g. "forwardedRef"
>   // And it can then be attached to the Component.
>   return React.forwardRef((props, ref) => {
>     return <LogProps {...props} forwardedRef={ref} />;
>   });
> ```

Sorry, my bad! I was just thinking about the concept while preparing this code snippet so I just took some parts of it from the official docs forgot to replace the names! 😅
